### ISSUE TITLE
Algorithm common interface

### ISSUE THREAD

**Issue by dav-m85** [graphp/algorithms#19]
When looking at the different algorithm providen, I cannot help feeling that they should share a common class interface.

When one look at Flow, one will find a few functions providing various outputs. However, class name does not help to know that one will find getBalance method wihtin. On the contrary, one get MinimumSpanningTree algorithms nicely sorted in a folder, with base class giving a common interface.

I would suggest rewrite algorithms by making them implement common interfaces, like this:
```php
interface GraphToVerticesAlgorithm {
public function compute(Graph $graph): Vertices;
}
interface GraphToBooleanAlgorithm {
public function compute(Graph $graph): Boolean;
}
class StronglyConnectedComponent\Tarjan implements GraphToVerticesAlgorithm {...}
class Eulerian\HasCycle implements GraphToVerticesAlgorithm {...}
```

I'll emphasize also on reusability of Algorithms. They shall not persist state between calls to compute function. You could reuse an Algorithm object multiple times.

I'm aware that this rewrite would be costly and maybe discouraging to even more contributors to port their pull requests. This is just a suggestion.

**Comment by clue** [comment]
> When one look at Flow, one will find a few functions providing various outputs. However, class name does not help to know that one will find getBalance method wihtin

I agree that the current class hierarchy is *suboptimal*, see also #5 and #4 for some background.

However, I'm not sure the suggested algorithm interfaces will help improve this.

Who is the consumer of these interfaces? How is this consumer going to benefit from this?

Take for example the `ShortestPath` namespace: Most consumers of these algorithms probably don't care about the concrete implementation (neglecting negative weights here). This means that if I am to write an algorithm which builds on top of shortest paths, it probably makes sense to inject (DI) the concrete algorithm implementation (likely with some conservative defaults). In this particular case, a common interface for these algorithms makes perfects sense. But a generic "AlgorithmThatReturnsEdges" interface would provide little value IMO.

> I'll emphasize also on reusability of Algorithms. They shall not persist state between calls to compute function. You could reuse an Algorithm object multiple times.

Agreed :-) Afaict this also already applies to all existing algorithms? We may want to explicitly state so somewhere in the documentation?

**Comment by phyrwork** [comment]
I can see how this might be useful. The interfaces could use tidying project-wide anyhow, even if just for consistency.

> ```
> interface GraphToVerticesAlgorithm {
> public function compute(Graph $graph): Vertices;
> }
> interface GraphToBooleanAlgorithm {
> public function compute(Graph $graph): Boolean;
> }
> class StronglyConnectedComponent\Tarjan implements GraphToVerticesAlgorithm {...}
> class Eulerian\HasCycle implements GraphToBooleanAlgorithm {...}
> ```

Alternatively

 ```
 interface VerticesAlgorithm {
 public function Vertices(Graph $graph): Vertices;
 }
 interface BoolAlgorithm {
 public function Bool(Graph $graph): Boolean;
 }
 class StronglyConnectedComponent\Tarjan implements VerticesAlgorithm {...}
 class Eulerian\HasCycle implements BoolAlgorithm {...}
 ```

to avoid namespacing collisions on special implementations that might satisfy both interfaces.

**Comment by clue** [comment]
@phyrwork I appreciate the discussion and agree that the current algorithm base interfaces are somewhat awkward and something we totally should address in one of the next releases!

That being said, do you think the suggested interfaces really address any of the concerns raised in my previous answer?

It's my understanding that interfaces should be exposed for *consumers* of this library to rely on.

Providing consistent interfaces across different classes of algorithms sounds appealing, but the way I see this is that is what caused part of the existing mess. Different classes of algorithms need different kind of interfaces.

My vote would be to remove the existing base classes for now and relax these API requirements in the upcoming releases. Once this is done we may reconsider applying consistent interfaces where it makes sense.
